File: 백준/Silver/20922. 겹치는 건 싫어/겹치는 건 싫어.py
import sys

input = sys.stdin.readline
n, k = map(int, input().split())
arr = list(map(int, input().split()))

# 각 숫자의 개수를 담는 배열
cnt = [0] * (max(arr) + 1)
left, right = 0, 0
answer = 0

while right < n:
   if cnt[arr[right]] < k:
      # cnt++ 후 오른쪽으로 전진
      cnt[arr[right]] += 1
      right += 1
   else:
      # 맨 왼쪽에서 한칸 오른쪽으로
      cnt[arr[left]] -= 1
      left += 1
   answer = max(answer, right - left)
   
print(answer)

File: 백준/Silver/20922. 겹치는 건 싫어/겹치는 건 싫어.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int[] arr = new int[N];
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        HashMap<Integer, Integer> map = new HashMap<>(); // 수, 겹침수
        int start = 0, end = 0, len = 0;
        int answer = 0;

        while (end < N){
            int now = arr[end];
            map.put(now, map.getOrDefault(now, 0) + 1);
            end++;

            // K 보다 크면 start 범위 줄이기
            while (map.get(now) > K){
                int s = arr[start];
                map.put(s, map.get(s) - 1);
                if (map.get(s) == 0) map.remove(s);
                start++;
            }

            answer = Math.max(answer, end - start);
        }

        bw.write(answer + "");
        bw.flush();
        bw.close();
    }
}